test/integration/auth.spec.ts

```diff
@@ -710,6 +710,49 @@ describe('admin.auth', () => {
             expect(userRecord.uid).to.equal(createdUserUid);
           });
       });
+
+      it('createCustomToken() mints a JWT that can be used to sign in tenant users', () => {
+        return tenantAwareAuth.createCustomToken('uid1')
+          .then((customToken) => {
+            firebase.auth().tenantId = createdTenantId;
+            return firebase.auth().signInWithCustomToken(customToken);
+          })
+          .then(({user}) => {
+            return user.getIdToken();
+          })
+          .then((idToken) => {
+            return tenantAwareAuth.verifyIdToken(idToken);
+          })
+          .then((token) => {
+            expect(token.uid).to.equal('uid1');
+            expect(token.firebase.tenant).to.equal(createdTenantId);
+          });
+      });
+
+      it('createCustomToken() should fail if tenantIds mismatch', async () => {
+        const aDifferentTenant = await admin.auth().tenantManager().createTenant({displayName: 'A-Different-Tenant'});
```

Comment:
Tenants might not get cleaned up if the test is interrupted. If you just want to test a failure case, you can just set a random tenantId in the custom token or in the sign in request without creating new tenant, and it should fail with a similar message

Reply:
I wanted to ensure that the mismatched tenant id was otherwise valid. For instance; you could imagine a backend implementation that inspected the tenant id to ensure it was valid, and *only then* checked the token to ensure it matched. By using an otherwise valid tenant id, I think this test becomes *slightly* stronger. (As a minor additional benefit, it also ensures I use a properly formatted tenant id; even in the unlikely event that the definition of a properly formatted tenant id changes in the future.)

It is true that if the test is interrupted, the tenant might not be cleaned up. But the existing code also suffers from this, and I'm not sure we can really make that bullet-proof. (The existing code creates a few tenants/users as well and then deletes them after the tests are completed.) We could just delete *all* users/tenants upon starting the tests... but this would prevent us from running two instances of the tests at once (though that probably doesn't work anyways) and would be dangerous if users accidentally supplied a "real" project to the test suite (well; more dangerous.)

Incidentally, as I was looking at this comment, I noticed that I was missing an await() in the cleanup code; fixed.

Reply:
But it seems the thing you are checking here is backend implementation. Does Admin sdk integration test need to cover that? Also it seems a bit weird to me that only this test needs to have await()? I will leave to @bojeil-google to make the judgement here.

Reply:
tl;dr: You're right; this test isn't necessary. I've removed it.



(The async() allows this to work without having to chain promises; which is especially useful when using try/finally. But you can get by just fine without it by either registering created tenants/etc and deleting them in the after() method, or by `.catch()`ing at the end of a promise chain. I think async() makes this more readable, but am happy to convert to either of the other approaches.)

This test really is testing the error handling on the sdk side; it behaves differently in the presence of tenant id mismatch vs invalid tenant id. (Details below.) Although, signInWithCustomToken is actually client sdk rather than admin sdk, so  this test should be moved there; you're exactly right that the admin sdk itself can't possibly care whether the tenant id is invalid or mismatched. I've removed this test.

Details: If I use an invalid tenant id (rather than a mismatching one) and then attempt to use the resulting token to login via the client sdk, I get this error: `The Auth instance's tenant ID is invalid.` with a code of `invalid-tenant-id`. But if I use a valid, but mismatching tenant id, I get `Specified tenant ID does not match the custom token.` with a code of `internal-error`. (Which is a bug in the client sdk; the code should be 'tenant-id-mismatch'; fixed here: https://github.com/firebase/firebase-js-sdk/pull/2389. In fact, that PR effectively performs this exact test.)

Reply:
Thanks for the quick fix on the client SDK bug!